When a single-node cluster reboots, mongo vanishes

Similarly to https://github.com/flynn/flynn/issues/3552 my MongoDB dies when I restart a single node cluster. I've an app requiring MongoDB and would really love to continue using Flynn. 
However, I think a node should withstand a reboot without having to rebuild the entire stack.

```flynn -a appname mongodb mongo``` results in the following:

``` MongoDB shell version: 3.2.10
connecting to: leader.mongodb.discoverd:27017/f1716ca9bbaa2ac8e3ced9c306f0226e
2016-10-25T06:15:32.412+0000 I NETWORK  [thread1] getaddrinfo("leader.mongodb.discoverd") failed: Name or service not known
2016-10-25T06:15:32.420+0000 E QUERY    [thread1] Error: couldn't initialize connection to host leader.mongodb.discoverd, address is invalid :
connect@src/mongo/shell/mongo.js:231:14
@(connect):1:6```

This is the collect-debug-info gist from immediately after the reboot: https://gist.github.com/6d445d26b52847cfc5aac7950bc99a2b

I've tried applying flynn-host fix -n 1. Unfortunately, some segmentation violation occurs upon that process:

```
root@ubuntu-2gb-fra1-01:~# flynn-host fix -n 1
INFO[10-25|06:18:59] found expected hosts                     n=1
INFO[10-25|06:18:59] ensuring discoverd is running on all hosts
INFO[10-25|06:19:00] checking flannel
INFO[10-25|06:19:00] flannel looks good
INFO[10-25|06:19:00] waiting for discoverd to be available
INFO[10-25|06:19:00] checking for running controller API
INFO[10-25|06:19:00] found running controller API instances   n=1
INFO[10-25|06:19:00] found controller instance, checking critical formations
INFO[10-25|06:19:00] checking status of sirenia databases
INFO[10-25|06:19:00] checking for database state              db=postgres
INFO[10-25|06:19:00] checking sirenia cluster status          fn=FixSirenia service=postgres
INFO[10-25|06:19:00] found running leader                     fn=FixSirenia service=postgres
INFO[10-25|06:19:00] found running instances                  fn=FixSirenia service=postgres count=1
INFO[10-25|06:19:00] getting sirenia status                   fn=FixSirenia service=postgres
INFO[10-25|06:19:00] cluster claims to be read-write          fn=FixSirenia service=postgres
INFO[10-25|06:19:00] checking for database state              db=mariadb
INFO[10-25|06:19:00] skipping recovery of db, no state in discoverd db=mariadb
INFO[10-25|06:19:00] checking for database state              db=mongodb
INFO[10-25|06:19:00] checking sirenia cluster status          fn=FixSirenia service=mongodb
INFO[10-25|06:19:00] no running leader                        fn=FixSirenia service=mongodb
INFO[10-25|06:19:00] found running instances                  fn=FixSirenia service=mongodb count=3
INFO[10-25|06:19:00] getting sirenia status                   fn=FixSirenia service=mongodb
INFO[10-25|06:19:00] getting service metadata                 fn=FixSirenia service=mongodb
INFO[10-25|06:19:00] getting primary job info                 fn=FixSirenia service=mongodb job.id=ubuntu2gbfra101-1b4397fc-163d-4a51-8523-f986dbab606b
INFO[10-25|06:19:00] terminating unassigned sirenia instances fn=FixSirenia service=mongodb
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x729021]

goroutine 1 [running]:
panic(0xb20da0, 0xc4200100a0)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/panic.go:500 +0x1a1
github.com/flynn/flynn/pkg/shutdown.(*handler).exit(0xc420015410, 0x0, 0x0, 0x0, 0xb20da0, 0xc4200100a0)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/pkg/shutdown/handler.go:69 +0x222
github.com/flynn/flynn/pkg/shutdown.Exit()
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/pkg/shutdown/handler.go:40 +0x6a
panic(0xb20da0, 0xc4200100a0)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/panic.go:458 +0x243
github.com/flynn/flynn/host/fixer.(*ClusterFixer).FixSirenia(0xc42016ec90, 0xbeb8ce, 0x7, 0x0, 0x2)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/host/fixer/sirenia.go:111 +0xbe1
github.com/flynn/flynn/host/fixer.(*ClusterFixer).Run(0xc42016ec90, 0xc42028b060, 0xc42028b0c0, 0x0, 0xc42028b0c0)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/host/fixer/fixer.go:122 +0x1089
github.com/flynn/flynn/host/fixer.(*ClusterFixer).Run-fm(0xc42028b060, 0xc42028b0c0, 0xc4200d7170, 0x3)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/host/cli/fix.go:8 +0x3e
github.com/flynn/flynn/host/cli.Run(0x7ffc633898cf, 0x3, 0xc420209ba0, 0x2, 0x2, 0xc4201d7c01, 0xc68ce8)
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/host/cli/cli.go:57 +0x3ca
main.main()
	/tmp/tmp.Qxrkc8u1kZ/src/github.com/flynn/flynn/host/host.go:149 +0x4d4
```


And this is the collect-debug-info gist AFTER I've tried to use ```flynn-host fix``` command: 
https://gist.github.com/4d8089d8eaed06dafd7643a0ef1a40c0

Maybe this helps to identify the issue? I've been able to reproduce it on my local VMWare machine and DigitalOcean instances...

In that other ticket, one reply states 

> I think I have identified the issue. It seems that there is a bug in the MongoDB appliance which >is only hit when both of the existing Sync and Primary peers go down and when resurrected >have different IP addresses.
>
>Thanks for reporting!

Which is also the same case here, if I am not mistaking?

--- a/appliance/mongodb/process.go
+++ b/appliance/mongodb/process.go
@@ -362,6 +362,9 @@ func (p *Process) assumePrimary(downstream *discoverd.Instance, clusterState *st
 		panic(fmt.Sprintf("unexpected state running role=%s", p.config().Role))
 	}
 
+	// Begin with both replication and security disabled
+	// We will enable both once we either initialise the database or confirm
+	// that it has already been initialized.
 	p.securityEnabledValue.Store(false)
 	if err := p.writeConfig(configData{}); err != nil {
 		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
@@ -437,8 +440,15 @@ func replSetGetStatusQuery(session *mgo.Session) (*replSetStatus, error) {
 func (p *Process) isReplInitialised() (bool, error) {
 	_, err := p.replSetGetStatus()
 	if err != nil {
-		if merr, ok := err.(*mgo.QueryError); ok && merr.Code == 94 {
-			return false, nil
+		if merr, ok := err.(*mgo.QueryError); ok {
+			switch merr.Code {
+			case 93: // replica set exists but is invalid/we aren't a member
+				return true, nil
+			case 94: // replica set not yet configured
+				return false, nil
+			}
+			p.Logger.Error("failed to check if replset initialized", "err", err, "code", merr.Code)
+			return false, err
 		}
 		return false, err
 	}
@@ -454,7 +464,7 @@ func (p *Process) isUserCreated() (bool, error) {
 
 	session.SetMode(mgo.Monotonic, true)
 
-	n, err := session.DB("system").C("users").Find("flynn").Count()
+	n, err := session.DB("admin").C("system.users").Find(bson.M{"user": "flynn"}).Count()
 	if err != nil {
 		if merr, ok := err.(*mgo.QueryError); ok && merr.Code == 13 {
 			return false, nil
@@ -495,9 +505,10 @@ func (p *Process) initPrimaryDB(clusterState *state.State) error {
 	logger.Info("initializing primary database")
 
 	// check if admin user has been created
+	logger.Info("checking if user has been created")
 	created, err := p.isUserCreated()
 	if err != nil {
-		logger.Error("error creating user")
+		logger.Error("error checking if user created")
 		return err
 	}
 
@@ -517,41 +528,48 @@ func (p *Process) initPrimaryDB(clusterState *state.State) error {
 		if err := p.start(); err != nil {
 			return err
 		}
+		logger.Info("creating user")
 		if err := p.createUser(); err != nil {
 			return err
 		}
-		logger.Info("stopping database to re-enable security")
-		if err := p.stop(); err != nil {
-			return err
-		}
-		p.securityEnabledValue.Store(true)
-		if err := p.writeConfig(configData{ReplicationEnabled: true}); err != nil {
-			logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
-			return err
-		}
-		logger.Info("starting database with security enabled")
-		if err := p.start(); err != nil {
-			return err
-		}
-		logger.Info("user created successfully")
+	}
+	logger.Info("stopping database to enable security/replication")
+	if err := p.stop(); err != nil {
+		return err
+	}
+	p.securityEnabledValue.Store(true)
+	if err := p.writeConfig(configData{ReplicationEnabled: true}); err != nil {
+		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+	logger.Info("starting database with security enabled")
+	if err := p.start(); err != nil {
+		return err
 	}
 
 	// check if replica set has been initialised
 	logger.Info("checking if replica set has been initialised")
 	initialized, err := p.isReplInitialised()
 	if err != nil {
+		logger.Error("error checking replset initialised", "err", err)
 		return err
 	}
+	logger.Info("not initialized, initialising now")
 	if !initialized && clusterState != nil {
 		if err := p.replSetInitiate(); err != nil {
+			logger.Error("error initialising replset", "err", err)
 			return err
 		}
 
 	}
+	logger.Info("getting current replset config")
 	replSetCurrent, err := p.getReplConfig()
 	if err != nil {
+		logger.Error("error getting replset config", "err", err)
 		return err
 	}
+
+	logger.Info("reconfiguring replset")
 	replSetNew := p.replSetConfigFromState(replSetCurrent, clusterState)
 	err = p.setReplConfig(replSetNew)
 	if err != nil {

--- a/host/fixer/fixer.go
+++ b/host/fixer/fixer.go
@@ -119,8 +119,13 @@ func (f *ClusterFixer) Run(args *docopt.Args, c *cluster.Client) error {
 			f.l.Error("error checking database state", "db", db)
 			return err
 		}
-		if err := f.FixSirenia(db); err != nil {
-			return err
+		if err := f.CheckSirenia(db); err != nil {
+			if err := f.KillSchedulers(); err != nil {
+				return err
+			}
+			if err := f.FixSirenia(db); err != nil {
+				return err
+			}
 		}
 	}
 

--- a/host/fixer/sirenia.go
+++ b/host/fixer/sirenia.go
@@ -108,7 +108,7 @@ func (f *ClusterFixer) FixSirenia(svc string) error {
 	log.Info("terminating unassigned sirenia instances")
 outer:
 	for _, i := range instances {
-		if i.Addr == state.Primary.Addr || i.Addr == state.Sync.Addr {
+		if i.Addr == state.Primary.Addr || (state.Sync != nil && i.Addr == state.Sync.Addr) {
 			continue
 		}
 		for _, a := range state.Async {

--- a/host/fixer/sirenia.go
+++ b/host/fixer/sirenia.go
@@ -13,8 +13,8 @@ import (
 	state "github.com/flynn/flynn/pkg/sirenia/state"
 )
 
-func (f *ClusterFixer) FixSirenia(svc string) error {
-	log := f.l.New("fn", "FixSirenia", "service", svc)
+func (f *ClusterFixer) CheckSirenia(svc string) error {
+	log := f.l.New("fn", "CheckSirenia", "service", svc)
 	log.Info("checking sirenia cluster status")
 	service := discoverd.NewService(svc)
 	leader, _ := service.Leader()
@@ -41,9 +41,18 @@ func (f *ClusterFixer) FixSirenia(svc string) error {
 		log.Info("cluster claims to be read-write")
 		return nil
 	}
+	return fmt.Errorf("cluster isn't read-write")
+}
+
+func (f *ClusterFixer) FixSirenia(svc string) error {
+	log := f.l.New("fn", "FixSirenia", "service", svc)
+
+	service := discoverd.NewService(svc)
+	instances, _ := service.Instances()
+	leader, _ := service.Leader()
 
 	log.Info("getting service metadata")
-	meta, err := discoverd.NewService(svc).GetMeta()
+	meta, err := service.GetMeta()
 	if err != nil {
 		return fmt.Errorf("error getting sirenia state from discoverd: %s", err)
 	}

--- a/host/fixer/sirenia.go
+++ b/host/fixer/sirenia.go
@@ -220,10 +220,10 @@ outer:
 		if err != nil {
 			return err
 		}
-		if leader != nil {
+		if leader != nil && leader.Addr != "" {
 			addr = leader.Addr
 		}
-		log.Info("waiting for cluster to come up read-write")
+		log.Info("waiting for cluster to come up read-write", "addr", addr)
 		return sirenia.NewClient(addr).WaitForReadWrite(5 * time.Minute)
 	}
 	return nil

--- a/host/monitor.go
+++ b/host/monitor.go
@@ -208,11 +208,13 @@ func (m *Monitor) repairCluster() error {
 			log.Error("error checking database state", "db", db)
 			return err
 		}
-		if err := f.FixSirenia(db); err != nil {
-			if db == "postgres" {
-				return err
-			} else {
-				log.Error("failed database recovery", "db", db)
+		if err := f.CheckSirenia(db); err != nil {
+			if err := f.FixSirenia(db); err != nil {
+				if db == "postgres" {
+					return err
+				} else {
+					log.Error("failed database recovery", "db", db)
+				}
 			}
 		}
 	}

